### PR TITLE
feat(pipeline): action-agent support edge pipeline

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
action-agent support edge pipeline

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=300237&iterationID=1174&pId=301655&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English | action-agent support edge pipeline             |
| 🇨🇳 中文    |     action-agent支持边缘pipeline上报         |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by chengjoey** [comment]
tested in dev, edge-pipeline agent doesn't support cms and upload files
Others are the same as normal pipeline
image

File: conf/pipeline/pipeline.yaml [hunk 1]
```diff
@@ -39,6 +39,7 @@ clusterinfo: {}
 edgepipeline_register:
   cluster_dialer_endpoint: "${CLUSTER_DIALER_PUBLIC_URL}/clusteragent/connect"
   cluster_access_key: "${CLUSTER_ACCESS_KEY}"
+  access_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
```
Comment (sfwn) [inline conf/pipeline/pipeline.yaml:42]: env configurable is required.

File: modules/actionagent/callback.go [hunk 2]
```diff
@@ -76,6 +75,28 @@ func (agent *Agent) Callback() {
 	}
 }
 
+func (agent *Agent) SetCallbackReporter() {
+	tokenForBootstrap := os.Getenv(apistructs.EnvOpenapiTokenForActionBootstrap)
+	if tokenForBootstrap == "" {
+		agent.AppendError(errors.Errorf("missing env %s", apistructs.EnvOpenapiTokenForActionBootstrap))
+		return
+	}
+	agent.EasyUse.TokenForBootstrap = tokenForBootstrap
```
Comment (sfwn) [inline modules/actionagent/callback.go:84]: split into two single methods is clearly.

File: modules/actionagent/callback.go [hunk 3]
```diff
@@ -89,9 +110,12 @@ func (agent *Agent) callbackToPipelinePlatform(cb *Callback) (err error) {
 		}
 	}()
 
-	if agent.EasyUse.OpenAPIAddr == "" {
+	if agent.EasyUse.OpenAPIAddr == "" && !agent.EasyUse.IsEdgePipeline {
```
Comment (sfwn) [inline modules/actionagent/callback.go:113]: agent.canDoEdgeCallback
agent.canDoNormalCallback

File: modules/actionagent/openapi.go [hunk 4]
```diff
@@ -36,21 +38,30 @@ func (agent *Agent) getOpenAPIInfo() {
...
 
 			// 判断是否是边缘集群
 			agent.isEdgeCluster()
+			agent.isEdgePipeline()
 
 			// 根据是否是边缘集群对 openapi 环境变量进行转换
 			agent.convertEnvsByClusterLocation()
 			openAPIAddr := os.Getenv(EnvDiceOpenapiAddr)
-			if openAPIAddr == "" {
+			if openAPIAddr == "" && !agent.EasyUse.IsEdgePipeline {
```
Comment (sfwn) [inline modules/actionagent/openapi.go:50]: abstract

File: modules/pipeline/conf/conf.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ type Conf struct {
 	ListenAddr      string `env:"LISTEN_ADDR" default:":3081"`
 	Debug           bool   `env:"DEBUG" default:"false"`
 	DiceClusterName string `env:"DICE_CLUSTER_NAME" default:"local"` // 服务所在集群
+	DiceIsEdge      bool   `env:"DICE_IS_EDGE" default:"false"`      // is edge cluster
```
Comment (sfwn) [inline modules/pipeline/conf/conf.go:30]: do not add new items in conf.
define it in a proper provider please.

File: modules/pipeline/endpoints/callback.go [hunk 6]
```diff
@@ -44,6 +45,13 @@ func (e *Endpoints) pipelineCallback(ctx context.Context, r *http.Request, vars
 		return errorresp.ErrResp(err)
 	}
 
+	if conf.DiceIsEdge() {
+		token := r.Header.Get("Authorization")
+		if err := e.edgeRegister.CheckAccessToken(token); err != nil {
```
Comment (sfwn) [inline modules/pipeline/endpoints/callback.go:50]: edgeRegister.CheckAccessToken
edgeRegister.CheckAccessTokenFromHttpRequest

declare both methods

File: modules/pipeline/providers/edgepipeline_register/interface.go [hunk 7]
```diff
@@ -27,9 +28,43 @@ import (
...
 
+func (p *provider) GetAccessToken(req apistructs.OAuth2TokenGetRequest) (*apistructs.OAuth2Token, error) {
+	return &apistructs.OAuth2Token{
+		AccessToken: p.Cfg.accessToken,
+		ExpiresIn:   0,
+		TokenType:   "Bearer",
+	}, nil
+}
+
+func (p *provider) GetOAuth2Token(req apistructs.OAuth2TokenGetRequest) (*apistructs.OAuth2Token, error) {
```
Comment (sfwn) [inline modules/pipeline/providers/edgepipeline_register/interface.go:46]: ... req is not used?

File: modules/pipeline/providers/edgepipeline_register/interface.go [hunk 8]
```diff
@@ -27,9 +28,43 @@ import (
...
 
 type Interface interface {
+	GetAccessToken(req apistructs.OAuth2TokenGetRequest) (*apistructs.OAuth2Token, error)
+	GetOAuth2Token(req apistructs.OAuth2TokenGetRequest) (*apistructs.OAuth2Token, error)
+	GetEdgePipelineEnvs() apistructs.ClusterDialerClientDetail
+	CheckAccessToken(token string) error
 	//RegisterEdgeToDialer(ctx context.Context)
 }
 
+func (p *provider) GetAccessToken(req apistructs.OAuth2TokenGetRequest) (*apistructs.OAuth2Token, error) {
```
Comment (sfwn) [inline modules/pipeline/providers/edgepipeline_register/interface.go:38]: ... req is not used?

File: modules/pipeline/providers/reconciler/taskrun/taskop/prepare.go [hunk 9]
```diff
@@ -354,6 +354,11 @@ func (pre *prepare) makeTaskRun() (needRetry bool, err error) {
 	task.Extra.PublicEnvs["PIPELINE_REQUESTED_MEM"] = fmt.Sprintf("%g", task.Extra.RuntimeResource.Memory)
 	task.Extra.PublicEnvs["PIPELINE_REQUESTED_DISK"] = fmt.Sprintf("%g", task.Extra.RuntimeResource.Disk)
 
+	// edge pipeline envs
+	edgePipelineEnvs := pre.EdgeRegister.GetEdgePipelineEnvs()
```
Comment (sfwn) [inline modules/pipeline/providers/reconciler/taskrun/taskop/prepare.go:358]: `GetEdgePipelineEnvs` should in edge_pipeline or edge_pipeline_register?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: conf/pipeline/pipeline.yaml (hunk 1 above)
Comment (chengjoey) [reply conf/pipeline/pipeline.yaml:42]: adjusted

File: modules/actionagent/callback.go (hunk 2 above)
Comment (chengjoey) [reply modules/actionagent/callback.go:84]: adjusted

File: modules/actionagent/callback.go (hunk 3 above)
Comment (chengjoey) [reply modules/actionagent/callback.go:113]: adjusted

File: modules/actionagent/openapi.go (hunk 4 above)
Comment (chengjoey) [reply modules/actionagent/openapi.go:50]: adjusted

File: modules/pipeline/endpoints/callback.go (hunk 6 above)
Comment (chengjoey) [reply modules/pipeline/endpoints/callback.go:50]: adjusted

File: modules/pipeline/providers/edgepipeline_register/interface.go (hunk 7 above)
Comment (chengjoey) [reply modules/pipeline/providers/edgepipeline_register/interface.go:46]: now, all edge pipeline, task will return the same token

File: modules/pipeline/providers/edgepipeline_register/interface.go (hunk 8 above)
Comment (chengjoey) [reply modules/pipeline/providers/edgepipeline_register/interface.go:38]: now, all edge pipeline, task will return the same token

File: modules/pipeline/providers/reconciler/taskrun/taskop/prepare.go (hunk 9 above)
Comment (chengjoey) [reply modules/pipeline/providers/reconciler/taskrun/taskop/prepare.go:358]: in the feature, should impl in env-provider

**Review by sfwn — APPROVED** [review]

**Pull request merged**